### DISCUSSION TITLE
`ftd.json`

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#1035] [Ideas & RFCs]
Status: We have implemented `ftd.json`:

```ftd
-- ftd.json:
html: $html
age: $age
foo: bar
something: 2
person: $person
```

This works, it generates JSON like:

```json
{
    "html": "<whatever is $html is>",
    "age":  23,
    "foo": "bar",
    "something": 2,
    "person": { "name": "name of person", "title": "Mr" }
}
```

You can also use `if` with `ftd.json`. 

NOTE: We are no longer planning to do what's in the rest of this discussion, and are going in the [type safe templating/json generation direction](https://github.com/orgs/fastn-stack/discussions/2064).

-----

Note: This is part of [backend using ftd](https://github.com/fastn-stack/fastn/discussions/1081).

We should allow

```ftd
-- ftd.json:

{
     "we": {
            "can": ["also", $do, { "something-else": $this }]
     }
}
```

in ftd documents. `ftd.json` like `ftd.document` can not be repeated in one document. If accept header is content/json we return `ftd.json` as JSON.

**Comment by harshdoesdev** [comment]
Is this a correct example?:

```
-- import: fastn/processors as pr

-- record User:
integer id:
string name:
integer age:

-- User user:
$processor$: pr.pg
id: 1

SELECT * FROM users WHERE id=$id;

-- ftd.json:
user: $user

{
  "id": $user.id,
  "name": $user.name,
  "age": $user.age
}
```

Output:
```json
{
  "id": 1,
  "name": "jack",
  "age": 20
 }
```

**Example: Multiple Users**

```
-- import: fastn/processors as pr

-- record User:
integer id:
string name:

-- User user:
$processor$: pr.pg
id: 1

SELECT * FROM users;

-- ftd.json:
user: $users

{
  "id": $user.id,
  "name": $user.name,
  "age": $user.age
}
```

This will output:

```json
[
  {
  "id": 1,
  "name": "jack",
  "age": 20
 },
 {
  "id": 2,
  "name": "jill",
  "age": 21
 }
]
```

**Reply by amitu** [reply]
The second example is wrong.

**Comment by amitu** [comment]
We should treat the body of ftd.json as f-script that evaluated to an object. 

F-script means body of functions in fastn. 

It should allow everything that is allowed in function bodies.

**Reply by amitu** [reply]
Also for f-script we want to take as much inspiration from Rust as possible.

**Comment by harshdoesdev** [comment]
**Updated Example: Multiple Users**

```
-- import: fastn/processors as pr

-- record User:
integer id:
string name:

-- User user:
$processor$: pr.pg
id: 1

SELECT * FROM users;

-- ftd.json:
for: $user in $users

{
  "id": $user.id,
  "name": $user.name,
  "age": $user.age
}
```

This will output:

```json
[
  {
  "id": 1,
  "name": "jack",
  "age": 20
 },
 {
  "id": 2,
  "name": "jill",
  "age": 21
 }
]
```

**Reply by amitu** [reply]
No. This is not general enough. What if you have two for loops, or loop with an if, and functions calls etc etc. Write some JS or Python code, that is what we want. Proper function body.
